modules/mod-deep-learning/ModelCard.cpp

```diff
@@ -0,0 +1,371 @@
+   }
+
+   std::string toLowerCase(const std::string &string)
+   {
+      return audacity::ToUTF8(wxString(string).Lower());
+   }
+
+   std::string tryGetString(const std::string &key, const Doc& doc)
+   {
+      validateExists(key, doc);
+      if(!doc[key.c_str()].IsString())
+         throwTypeError(key, "string", doc);
+
+      return doc[key.c_str()].GetString();
+   }
+
+   std::string tryGetString(const std::string &key, const Doc& doc, const std::string &defaultValue)
+   {
+      try
+      {
+         return tryGetString(key, doc);
+      }
+      catch (const InvalidModelCardDocument &e)
+      {
+         return defaultValue;
+      }
+   }
+
+   int tryGetInt(const std::string &key, const Doc& doc)
+   {
+      validateExists(key, doc);
+      if(!doc[key.c_str()].IsInt())
+         throwTypeError(key, "int", doc);
+      
+      return doc[key.c_str()].GetInt();
+   }
+
+   int tryGetInt(const std::string &key, const Doc& doc, int defaultValue)
+   {
+      try
+      {
+         return tryGetInt(key, doc);
+      }
+      catch (const InvalidModelCardDocument &e)
+      {
+         return defaultValue;
+      }
+   }
+
+   bool tryGetBool(const std::string &key, const Doc& doc)
+   {
+      validateExists(key, doc);
+      if(!doc[key.c_str()].IsBool())
+         throwTypeError(key, "bool", doc);
+      
+      return doc[key.c_str()].GetBool();
+   }
+
+   bool tryGetBool(const std::string &key, const Doc& doc, bool defaultValue)
+   {
+      try
+      {
+         return tryGetBool(key, doc);
+      }
+      catch (const InvalidModelCardDocument &e)
+      {
+         return defaultValue;
+      }
+   }
+
+   std::vector<std::string> tryGetStringArray(const std::string &key, const Doc& doc, bool lower)
+   {
+      validateExists(key, doc);
+      if(!doc[key.c_str()].IsArray())
+         throwTypeError(key, "array", doc);
+      
+      std::vector<std::string> labels;
+      for (auto itr = doc[key.c_str()].Begin(); itr != doc[key.c_str()].End();  ++itr)
+      {
+         if (lower)
+            labels.emplace_back(toLowerCase(itr->GetString()));
+         else
+            labels.emplace_back(itr->GetString());
+      }
+
+      return labels;
+   }
+
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc)
+   {
+      validateExists(key, doc);
+      if(!doc[key.c_str()].IsUint64())
+         throwTypeError(key, "uint64", doc);
+      
+      return doc[key.c_str()].GetUint64();
+   }
+
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc, uint64_t defaultValue)
+   {
+      try
+      {
+         return tryGetUint64(key, doc);
+      }
+      catch (const InvalidModelCardDocument &e)
+      {
+         return defaultValue;
+      }
+   }
+}
+
+namespace parsers 
+{
+   Doc ParseString(const std::string &data)
+   {
+      Doc d;
+      // parse the data
+      d.Parse(data.c_str());
+      if (d.Parse(data.c_str()).HasParseError()) 
```

Comment:
`d.Parse` is called twice.  Is that a mistake?  Is it a harmless mistake that makes no difference to the final state of `d`?  I don't know rapidjson enough to answer the questions for myself.

Reply:
539a3ead07f5a298c8bca75203e772f8f74321a9